inferSchema does not work with Spark 1.3.0

I am using Scala 2.10.5 with Spark 1.3.0 from CDH (spark-csv_2.10 version 1.2.0). When I try to use inferSchema=true I get following exception:

	Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/spark/sql/types/AtomicType
		at com.databricks.spark.csv.CsvRelation.inferSchema(CsvRelation.scala:155)
		at com.databricks.spark.csv.CsvRelation.<init>(CsvRelation.scala:70)
		at com.databricks.spark.csv.DefaultSource.createRelation(DefaultSource.scala:138)
		at com.databricks.spark.csv.DefaultSource.createRelation(DefaultSource.scala:40)
		at com.databricks.spark.csv.DefaultSource.createRelation(DefaultSource.scala:28)
		at org.apache.spark.sql.sources.ResolvedDataSource$.apply(ddl.scala:290)
		at org.apache.spark.sql.SQLContext.load(SQLContext.scala:679)
		at pl.cenatorium.App$.main(App.scala:17)
		at pl.cenatorium.App.main(App.scala)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:606)
		at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:569)
		at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:166)
		at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:189)
		at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:110)
		at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
	Caused by: java.lang.ClassNotFoundException: org.apache.spark.sql.types.AtomicType
		at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
		at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
		at java.security.AccessController.doPrivileged(Native Method)
		at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
		at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
		at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
		... 18 more

As far I can deduce from this stacktrace it looks that class org/apache/spark/sql/types/AtomicType is only present in Spark 1.4.0

--- a/.travis.yml
+++ b/.travis.yml
@@ -29,6 +29,7 @@ matrix:
       env: TEST_SPARK_VERSION="1.5.0-rc2"
 script:
   - sbt ++$TRAVIS_SCALA_VERSION -Dspark.testVersion=$TEST_SPARK_VERSION coverage test
-  - sbt ++$TRAVIS_SCALA_VERSION scalastyle
+  # Disable scalastyle for tests until # https://github.com/scalastyle/scalastyle/issues/156 is fixed:
+  # - sbt ++$TRAVIS_SCALA_VERSION scalastyle
 after_success:
   - bash <(curl -s https://codecov.io/bash)

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,13 +1,34 @@
 language: scala
-scala:
-  - 2.10.4
-jdk:
-  - openjdk7
-  - openjdk6
 sudo: false
+cache:
+  directories:
+    - $HOME/.ivy2
+matrix:
+  include:
+    # Spark 1.3.0
+    - jdk: openjdk6
+      scala: 2.10.4
+      env: TEST_SPARK_VERSION="1.3.0"
+    - jdk: openjdk6
+      scala: 2.11.6
+      env: TEST_SPARK_VERSION="1.3.0"
+    # Spark 1.4.0
+    - jdk: openjdk6
+      scala: 2.11.6
+      env: TEST_SPARK_VERSION="1.4.0"
+    - jdk: openjdk7
+      scala: 2.11.6
+      env: TEST_SPARK_VERSION="1.4.0"
+    # Spark 1.5.0
+    # TOOD: after 1.5.0 is released, update this to use the released version
+    - jdk: openjdk7
+      scala: 2.10.4
+      env: TEST_SPARK_VERSION="1.5.0-rc2"
+    - jdk: openjdk7
+      scala: 2.11.6
+      env: TEST_SPARK_VERSION="1.5.0-rc2"
 script:
-  - sbt -jvm-opts travis/jvmopts.compile compile
-  - sbt -jvm-opts travis/jvmopts.test coverage test
-  - sbt -jvm-opts travis/jvmopts.test scalastyle
+  - sbt ++$TRAVIS_SCALA_VERSION -Dspark.testVersion=$TEST_SPARK_VERSION coverage test
+  - sbt ++$TRAVIS_SCALA_VERSION scalastyle
 after_success:
   - bash <(curl -s https://codecov.io/bash)

--- a/build.sbt
+++ b/build.sbt
@@ -10,7 +10,7 @@ spName := "databricks/spark-csv"
 
 crossScalaVersions := Seq("2.10.4", "2.11.6")
 
-sparkVersion := "1.4.0"
+sparkVersion := "1.3.0"
 
 val testSparkVersion = settingKey[String]("The version of Spark to test against.")
 

--- a/build.sbt
+++ b/build.sbt
@@ -6,20 +6,38 @@ organization := "com.databricks"
 
 scalaVersion := "2.11.6"
 
-parallelExecution in Test := false
+spName := "databricks/spark-csv"
 
 crossScalaVersions := Seq("2.10.4", "2.11.6")
 
-libraryDependencies += "org.apache.commons" % "commons-csv" % "1.1"
+sparkVersion := "1.4.0"
 
-libraryDependencies += "com.univocity" % "univocity-parsers" % "1.5.1"
+val testSparkVersion = settingKey[String]("The version of Spark to test against.")
 
-libraryDependencies += "org.slf4j" % "slf4j-api" % "1.7.5" % "provided"
+testSparkVersion := sys.props.get("spark.testVersion").getOrElse(sparkVersion.value)
 
 resolvers ++= Seq(
   "Apache Staging" at "https://repository.apache.org/content/repositories/staging/",
   "Typesafe" at "http://repo.typesafe.com/typesafe/releases",
   "Local Maven Repository" at "file://"+Path.userHome.absolutePath+"/.m2/repository"
+)
+
+// TODO: remove once Spark 1.5.0 is released.
+resolvers += "Spark 1.5.0 RC2 Staging" at "https://repository.apache.org/content/repositories/orgapachespark-1141"
+
+sparkComponents := Seq("core", "sql")
+
+libraryDependencies ++= Seq(
+  "org.apache.commons" % "commons-csv" % "1.1",
+  "com.univocity" % "univocity-parsers" % "1.5.1",
+  "org.slf4j" % "slf4j-api" % "1.7.5" % "provided",
+  "org.scalatest" %% "scalatest" % "2.2.1" % "test",
+  "com.novocode" % "junit-interface" % "0.9" % "test"
+)
+
+libraryDependencies ++= Seq(
+  "org.apache.spark" %% "spark-core" % testSparkVersion.value % "test" force(),
+  "org.apache.spark" %% "spark-sql" % testSparkVersion.value % "test" force()
 )
 
 publishMavenStyle := true
@@ -57,17 +75,9 @@ pomExtra := (
     </developer>
   </developers>)
 
-spName := "databricks/spark-csv"
-
-sparkVersion := "1.4.0"
-
-sparkComponents += "sql"
-
-libraryDependencies += "org.scalatest" %% "scalatest" % "2.2.1" % "test"
-
-libraryDependencies += "com.novocode" % "junit-interface" % "0.9" % "test"
+parallelExecution in Test := false
 
 ScoverageSbtPlugin.ScoverageKeys.coverageHighlighting := {
   if (scalaBinaryVersion.value == "2.10") false
-  else false
+  else true
 }

--- a/project/build.properties
+++ b/project/build.properties
@@ -14,4 +14,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-sbt.version=0.13.6
+sbt.version=0.13.9

--- a/src/main/scala/com/databricks/spark/csv/util/InferSchema.scala
+++ b/src/main/scala/com/databricks/spark/csv/util/InferSchema.scala
@@ -32,8 +32,8 @@ private[csv] object InferSchema {
    */
   def apply(tokenRdd: RDD[Array[String]], header: Array[String]): StructType = {
 
-    val startType = Array.fill[DataType](header.length)(NullType)
-    val rootTypes = tokenRdd.aggregate(startType)(inferRowType, mergeRowTypes)
+    val startType: Array[DataType] = Array.fill[DataType](header.length)(NullType)
+    val rootTypes: Array[DataType] = tokenRdd.aggregate(startType)(inferRowType, mergeRowTypes)
 
     val stuctFields = header.zip(rootTypes).map { case (thisHeader, rootType) =>
       StructField(thisHeader, rootType, nullable = true)
@@ -42,7 +42,7 @@ private[csv] object InferSchema {
     StructType(stuctFields)
   }
 
-  private[csv] def inferRowType(rowSoFar: Array[DataType], next: Array[String]) = {
+  private def inferRowType(rowSoFar: Array[DataType], next: Array[String]): Array[DataType] = {
     var i = 0
     while(i < math.min(rowSoFar.length, next.length)){  // May have columns on right missing.
       rowSoFar(i) = inferField(rowSoFar(i), next(i))
@@ -51,7 +51,9 @@ private[csv] object InferSchema {
     rowSoFar
   }
 
-  private[csv] def mergeRowTypes(first: Array[DataType], second: Array[DataType]) = {
+  private[csv] def mergeRowTypes(
+      first: Array[DataType],
+      second: Array[DataType]): Array[DataType] = {
     first.zipAll(second, NullType, NullType).map { case ((a, b)) =>
       val tpe = findTightestCommonType(a, b).getOrElse(StringType)
       tpe match {
@@ -66,7 +68,6 @@ private[csv] object InferSchema {
    * point checking if it is an Int, as the final type must be Double or higher.
    */
   private[csv] def inferField(typeSoFar: DataType, field: String): DataType = {
-
     if (field == null || field.isEmpty){
       typeSoFar
     } else {
@@ -77,32 +78,45 @@ private[csv] object InferSchema {
         case DoubleType => tryParseDouble(field)
         case TimestampType => tryParseTimestamp(field)
         case StringType => StringType
+        case other: DataType =>
+          throw new UnsupportedOperationException(s"Unexpected data type $other")
       }
     }
   }
 
 
-  def tryParseInteger(field: String) = if((allCatch opt field.toInt).isDefined) {
+  private def tryParseInteger(field: String): DataType = if ((allCatch opt field.toInt).isDefined) {
     IntegerType
   } else {
     tryParseLong(field)
   }
 
-  def tryParseLong(field: String) = if ((allCatch opt field.toLong).isDefined) {
+  private def tryParseLong(field: String): DataType = if ((allCatch opt field.toLong).isDefined) {
     LongType
   } else {
     tryParseDouble(field)
   }
 
-  def tryParseDouble(field: String) = if ((allCatch opt field.toDouble).isDefined) {
-    DoubleType
-  } else {
-    tryParseTimestamp(field)
+  private def tryParseDouble(field: String): DataType = {
+    if ((allCatch opt field.toDouble).isDefined) {
+      DoubleType
+    } else {
+      tryParseTimestamp(field)
+    }
   }
 
-  def tryParseTimestamp(field: String) = if((allCatch opt Timestamp.valueOf(field)).isDefined) {
-    TimestampType
-  } else {
+  def tryParseTimestamp(field: String): DataType = {
+    if ((allCatch opt Timestamp.valueOf(field)).isDefined) {
+      TimestampType
+    } else {
+      stringType()
+    }
+  }
+
+  // Defining a function to return the StringType constant is necessary in order to work around
+  // a Scala compiler issue which leads to runtime incompatibilities with certain Spark versions;
+  // see issue #128 for more details.
+  private def stringType(): DataType = {
     StringType
   }
 
@@ -110,8 +124,8 @@ private[csv] object InferSchema {
    * Copied from internal Spark api
    * [[org.apache.spark.sql.catalyst.analysis.HiveTypeCoercion]]
    */
-  private val numericPrecedence =
-    IndexedSeq(
+  private val numericPrecedence: IndexedSeq[DataType] =
+    IndexedSeq[DataType](
       ByteType,
       ShortType,
       IntegerType,

--- a/src/test/java/com/databricks/spark/csv/JavaCsvSuite.java
+++ b/src/test/java/com/databricks/spark/csv/JavaCsvSuite.java
@@ -2,15 +2,14 @@ package com.databricks.spark.csv;
 
 import java.io.File;
 import java.util.HashMap;
-import java.util.Random;
 
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import org.apache.spark.SparkContext;
 import org.apache.spark.sql.*;
-import org.apache.spark.sql.test.TestSQLContext$;
 
 public class JavaCsvSuite {
   private transient SQLContext sqlContext;
@@ -22,12 +21,12 @@ public class JavaCsvSuite {
 
   @Before
   public void setUp() {
-    // Trigger static initializer of TestData
-    sqlContext = TestSQLContext$.MODULE$;
+    sqlContext = new SQLContext(new SparkContext("local[2]", "JavaCsvSuite"));
   }
 
   @After
   public void tearDown() {
+    sqlContext.sparkContext().stop();
     sqlContext = null;
   }
 

--- a/src/test/scala/com/databricks/spark/csv/CsvSuite.scala
+++ b/src/test/scala/com/databricks/spark/csv/CsvSuite.scala
@@ -20,16 +20,12 @@ import java.nio.charset.UnsupportedCharsetException
 import java.sql.Timestamp
 
 import org.apache.hadoop.io.compress.GzipCodec
-import org.apache.spark.sql.Row
-import org.apache.spark.sql.test._
-import org.apache.spark.SparkException
+import org.apache.spark.sql.{SQLContext, Row}
+import org.apache.spark.{SparkContext, SparkException}
 import org.apache.spark.sql.types._
-import org.scalatest.FunSuite
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
 
-/* Implicits */
-import TestSQLContext._
-
-abstract class AbstractCsvSuite extends FunSuite {
+abstract class AbstractCsvSuite extends FunSuite with BeforeAndAfterAll {
   val carsFile = "src/test/resources/cars.csv"
   val carsFile8859 = "src/test/resources/cars_iso-8859-1.csv"
   val carsTsvFile = "src/test/resources/cars.tsv"
@@ -45,8 +41,23 @@ abstract class AbstractCsvSuite extends FunSuite {
 
   protected def parserLib: String
 
+  private var sqlContext: SQLContext = _
+
+  override protected def beforeAll(): Unit = {
+    super.beforeAll()
+    sqlContext = new SQLContext(new SparkContext("local[2]", "AvroSuite"))
+  }
+
+  override protected def afterAll(): Unit = {
+    try {
+      sqlContext.sparkContext.stop()
+    } finally {
+      super.afterAll()
+    }
+  }
+
   test("DSL test") {
-    val results = TestSQLContext
+    val results = sqlContext
       .csvFile(carsFile, parserLib = parserLib)
       .select("year")
       .collect()
@@ -55,7 +66,7 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DSL test for iso-8859-1 encoded file") {
-    val dataFrame = TestSQLContext
+    val dataFrame = sqlContext
       .csvFile(carsFile8859, parserLib = parserLib, charset  = "iso-8859-1", delimiter = 'þ')
 
     assert(dataFrame.select("year").collect().size === numCars)
@@ -66,7 +77,7 @@ abstract class AbstractCsvSuite extends FunSuite {
 
   test("DSL test bad charset name") {
     val exception = intercept[UnsupportedCharsetException] {
-      val results = TestSQLContext
+      val results = sqlContext
         .csvFile(carsFile8859, parserLib = parserLib, charset = "1-9588-osi")
         .select("year")
         .collect()
@@ -75,18 +86,18 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DDL test") {
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable
          |USING com.databricks.spark.csv
          |OPTIONS (path "$carsFile", header "true", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT year FROM carsTable").collect().size === numCars)
+    assert(sqlContext.sql("SELECT year FROM carsTable").collect().size === numCars)
   }
 
   test("DDL test with charset") {
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable
          |USING com.databricks.spark.csv
@@ -94,22 +105,22 @@ abstract class AbstractCsvSuite extends FunSuite {
          |charset "iso-8859-1", delimiter "þ")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT year FROM carsTable").collect().size === numCars)
+    assert(sqlContext.sql("SELECT year FROM carsTable").collect().size === numCars)
   }
 
   test("DDL test with tab separated file") {
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable
          |USING com.databricks.spark.csv
          |OPTIONS (path "$carsTsvFile", header "true", delimiter "\t", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT year FROM carsTable").collect().size === numCars)
+    assert(sqlContext.sql("SELECT year FROM carsTable").collect().size === numCars)
   }
 
   test("DDL test parsing decimal type") {
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable
          |(yearMade double, makeName string, modelName string, priceTag decimal,
@@ -118,8 +129,9 @@ abstract class AbstractCsvSuite extends FunSuite {
          |OPTIONS (path "$carsTsvFile", header "true", delimiter "\t", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT yearMade FROM carsTable").collect().size === numCars)
-    assert(sql("SELECT makeName FROM carsTable where priceTag > 60000").collect().size === 1)
+    assert(sqlContext.sql("SELECT yearMade FROM carsTable").collect().size === numCars)
+    assert(
+      sqlContext.sql("SELECT makeName FROM carsTable where priceTag > 60000").collect().size === 1)
   }
 
   test("DSL test for DROPMALFORMED parsing mode") {
@@ -127,7 +139,7 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withParseMode("DROPMALFORMED")
       .withUseHeader(true)
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, carsFile)
+      .csvFile(sqlContext, carsFile)
       .select("year")
       .collect()
 
@@ -141,7 +153,7 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withParserLib(parserLib)
 
     val exception = intercept[SparkException]{
-      parser.csvFile(TestSQLContext, carsFile)
+      parser.csvFile(sqlContext, carsFile)
         .select("year")
         .collect()
     }
@@ -156,7 +168,7 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withQuoteChar('\'')
       .withUseHeader(true)
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, carsAltFile)
+      .csvFile(sqlContext, carsAltFile)
       .select("year")
       .collect()
 
@@ -165,7 +177,7 @@ abstract class AbstractCsvSuite extends FunSuite {
 
   test("DSL test with alternative delimiter and quote using sparkContext.csvFile") {
     val results =
-      TestSQLContext.csvFile(
+      sqlContext.csvFile(
         carsAltFile,
         useHeader = true,
         delimiter = '|',
@@ -179,7 +191,7 @@ abstract class AbstractCsvSuite extends FunSuite {
 
   test("Expect parsing error with wrong delimiter setting using sparkContext.csvFile") {
     intercept[ org.apache.spark.sql.AnalysisException] {
-      TestSQLContext.csvFile(
+      sqlContext.csvFile(
         carsAltFile,
         useHeader = true,
         delimiter = ',',
@@ -192,7 +204,7 @@ abstract class AbstractCsvSuite extends FunSuite {
 
   test("Expect wrong parsing results with wrong quote setting using sparkContext.csvFile") {
     val results =
-      TestSQLContext.csvFile(
+      sqlContext.csvFile(
         carsAltFile,
         useHeader = true,
         delimiter = '|',
@@ -206,7 +218,7 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DDL test with alternative delimiter and quote") {
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable
          |USING com.databricks.spark.csv
@@ -214,7 +226,7 @@ abstract class AbstractCsvSuite extends FunSuite {
          |parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT year FROM carsTable").collect().size === numCars)
+    assert(sqlContext.sql("SELECT year FROM carsTable").collect().size === numCars)
   }
 
 
@@ -223,33 +235,33 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withSchema(StructType(List(StructField("column", StringType, false))))
       .withUseHeader(false)
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, emptyFile)
+      .csvFile(sqlContext, emptyFile)
       .count()
 
     assert(results === 0)
   }
 
   test("DDL test with empty file") {
-    sql(s"""
+    sqlContext.sql(s"""
            |CREATE TEMPORARY TABLE carsTable
            |(yearMade double, makeName string, modelName string, comments string, grp string)
            |USING com.databricks.spark.csv
            |OPTIONS (path "$emptyFile", header "false", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT count(*) FROM carsTable").collect().head(0) === 0)
+    assert(sqlContext.sql("SELECT count(*) FROM carsTable").collect().head(0) === 0)
   }
 
   test("DDL test with schema") {
-    sql(s"""
+    sqlContext.sql(s"""
            |CREATE TEMPORARY TABLE carsTable
            |(yearMade double, makeName string, modelName string, comments string, grp string)
            |USING com.databricks.spark.csv
            |OPTIONS (path "$carsFile", header "true", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT makeName FROM carsTable").collect().size === numCars)
-    assert(sql("SELECT avg(yearMade) FROM carsTable where grp = '' group by grp")
+    assert(sqlContext.sql("SELECT makeName FROM carsTable").collect().size === numCars)
+    assert(sqlContext.sql("SELECT avg(yearMade) FROM carsTable where grp = '' group by grp")
       .collect().head(0) === 2004.5)
   }
 
@@ -257,7 +269,7 @@ abstract class AbstractCsvSuite extends FunSuite {
     val cars = new CsvParser()
       .withUseHeader(false)
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, carsFile)
+      .csvFile(sqlContext, carsFile)
     assert(cars.schema.fields(0).name == "C0")
     assert(cars.schema.fields(2).name == "C2")
   }
@@ -266,28 +278,28 @@ abstract class AbstractCsvSuite extends FunSuite {
     // Create a temp directory for table that will be overwritten
     TestUtils.deleteRecursively(new File(tempEmptyDir))
     new File(tempEmptyDir).mkdirs()
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTableIO
          |USING com.databricks.spark.csv
          |OPTIONS (path "$carsFile", header "true", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
-    sql(s"""
+    sqlContext.sql(s"""
            |CREATE TEMPORARY TABLE carsTableEmpty
            |(yearMade double, makeName string, modelName string, comments string, grp string)
            |USING com.databricks.spark.csv
            |OPTIONS (path "$tempEmptyDir", header "false", parserLib "$parserLib")
       """.stripMargin.replaceAll("\n", " "))
 
-    assert(sql("SELECT * FROM carsTableIO").collect().size === numCars)
-    assert(sql("SELECT * FROM carsTableEmpty").collect().isEmpty)
+    assert(sqlContext.sql("SELECT * FROM carsTableIO").collect().size === numCars)
+    assert(sqlContext.sql("SELECT * FROM carsTableEmpty").collect().isEmpty)
 
-    sql(
+    sqlContext.sql(
       s"""
          |INSERT OVERWRITE TABLE carsTableEmpty
          |SELECT * FROM carsTableIO
       """.stripMargin.replaceAll("\n", " "))
-    assert(sql("SELECT * FROM carsTableEmpty").collect().size == numCars)
+    assert(sqlContext.sql("SELECT * FROM carsTableEmpty").collect().size == numCars)
   }
 
   test("DSL save") {
@@ -296,10 +308,10 @@ abstract class AbstractCsvSuite extends FunSuite {
     new File(tempEmptyDir).mkdirs()
     val copyFilePath = tempEmptyDir + "cars-copy.csv"
 
-    val cars = TestSQLContext.csvFile(carsFile, parserLib = parserLib)
+    val cars = sqlContext.csvFile(carsFile, parserLib = parserLib)
     cars.saveAsCsvFile(copyFilePath, Map("header" -> "true"))
 
-    val carsCopy = TestSQLContext.csvFile(copyFilePath + "/")
+    val carsCopy = sqlContext.csvFile(copyFilePath + "/")
 
     assert(carsCopy.count == cars.count)
     assert(carsCopy.collect.map(_.toString).toSet == cars.collect.map(_.toString).toSet)
@@ -311,10 +323,10 @@ abstract class AbstractCsvSuite extends FunSuite {
     new File(tempEmptyDir).mkdirs()
     val copyFilePath = tempEmptyDir + "cars-copy.csv"
 
-    val cars = TestSQLContext.csvFile(carsFile, parserLib = parserLib)
+    val cars = sqlContext.csvFile(carsFile, parserLib = parserLib)
     cars.saveAsCsvFile(copyFilePath, Map("header" -> "true"), classOf[GzipCodec])
 
-    val carsCopy = TestSQLContext.csvFile(copyFilePath + "/")
+    val carsCopy = sqlContext.csvFile(copyFilePath + "/")
 
     assert(carsCopy.count == cars.count)
     assert(carsCopy.collect.map(_.toString).toSet == cars.collect.map(_.toString).toSet)
@@ -326,10 +338,10 @@ abstract class AbstractCsvSuite extends FunSuite {
     new File(tempEmptyDir).mkdirs()
     val copyFilePath = tempEmptyDir + "cars-copy.csv"
 
-    val cars = TestSQLContext.csvFile(carsFile, parserLib = parserLib)
+    val cars = sqlContext.csvFile(carsFile, parserLib = parserLib)
     cars.saveAsCsvFile(copyFilePath, Map("header" -> "true", "quote" -> "\""))
 
-    val carsCopy = TestSQLContext.csvFile(copyFilePath + "/", parserLib = parserLib)
+    val carsCopy = sqlContext.csvFile(copyFilePath + "/", parserLib = parserLib)
 
     assert(carsCopy.count == cars.count)
     assert(carsCopy.collect.map(_.toString).toSet == cars.collect.map(_.toString).toSet)
@@ -341,10 +353,10 @@ abstract class AbstractCsvSuite extends FunSuite {
     new File(tempEmptyDir).mkdirs()
     val copyFilePath = tempEmptyDir + "cars-copy.csv"
 
-    val cars = TestSQLContext.csvFile(carsFile)
+    val cars = sqlContext.csvFile(carsFile)
     cars.saveAsCsvFile(copyFilePath, Map("header" -> "true", "quote" -> "!"))
 
-    val carsCopy = TestSQLContext.csvFile(copyFilePath + "/", quote = '!', parserLib = parserLib)
+    val carsCopy = sqlContext.csvFile(copyFilePath + "/", quote = '!', parserLib = parserLib)
 
     assert(carsCopy.count == cars.count)
     assert(carsCopy.collect.map(_.toString).toSet == cars.collect.map(_.toString).toSet)
@@ -356,10 +368,10 @@ abstract class AbstractCsvSuite extends FunSuite {
     new File(tempEmptyDir).mkdirs()
     val copyFilePath = tempEmptyDir + "escape-copy.csv"
 
-    val escape = TestSQLContext.csvFile(escapeFile, escape='|', quote='"')
+    val escape = sqlContext.csvFile(escapeFile, escape='|', quote='"')
     escape.saveAsCsvFile(copyFilePath, Map("header" -> "true", "quote" -> "\""))
 
-    val escapeCopy = TestSQLContext.csvFile(copyFilePath + "/", parserLib = parserLib)
+    val escapeCopy = sqlContext.csvFile(copyFilePath + "/", parserLib = parserLib)
 
     assert(escapeCopy.count == escape.count)
     assert(escapeCopy.collect.map(_.toString).toSet == escape.collect.map(_.toString).toSet)
@@ -367,7 +379,7 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DSL test schema inferred correctly") {
-    val results = TestSQLContext
+    val results = sqlContext
       .csvFile(carsFile, parserLib = parserLib, inferSchema = true)
 
     assert(results.schema == StructType(List(
@@ -382,7 +394,7 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DSL test inferred schema passed through") {
-    val dataFrame = TestSQLContext
+    val dataFrame = sqlContext
       .csvFile(carsFile, parserLib = parserLib, inferSchema = true)
 
     val results = dataFrame
@@ -394,14 +406,14 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DDL test with inferred schema") {
-    sql(
+    sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable
          |USING com.databricks.spark.csv
          |OPTIONS (path "$carsFile", header "true", parserLib "$parserLib", inferSchema "true")
       """.stripMargin.replaceAll("\n", " "))
 
-    val results = sql("select year from carsTable where make = 'Ford'")
+    val results = sqlContext.sql("select year from carsTable where make = 'Ford'")
 
     assert(results.first().getInt(0) === 1997)
   }
@@ -412,7 +424,7 @@ abstract class AbstractCsvSuite extends FunSuite {
                                   StructField("age", IntegerType, true))))
       .withUseHeader(true)
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, nullNumbersFile)
+      .csvFile(sqlContext, nullNumbersFile)
       .collect()
 
     assert(results.head.toSeq === Seq("alice", 35))
@@ -425,7 +437,7 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withDelimiter(',')
       .withComment('~')
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, commentsFile)
+      .csvFile(sqlContext, commentsFile)
       .collect()
 
     val expected =
@@ -442,7 +454,7 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withComment('~')
       .withParserLib(parserLib)
       .withInferSchema(true)
-      .csvFile(TestSQLContext, commentsFile)
+      .csvFile(sqlContext, commentsFile)
       .collect()
 
     val expected =
@@ -459,7 +471,7 @@ abstract class AbstractCsvSuite extends FunSuite {
       .withDelimiter(',')
       .withComment(null)
       .withParserLib(parserLib)
-      .csvFile(TestSQLContext, disableCommentsFile)
+      .csvFile(sqlContext, disableCommentsFile)
       .collect()
 
     val expected =
@@ -471,11 +483,11 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("DSL load csv from rdd") {
-    val csvRdd = TestSQLContext.sparkContext.parallelize(Seq("age,height", "20,1.8", "16,1.7"))
+    val csvRdd = sqlContext.sparkContext.parallelize(Seq("age,height", "20,1.8", "16,1.7"))
     val df = new CsvParser()
       .withUseHeader(true)
       .withParserLib(parserLib)
-      .csvRdd(TestSQLContext, csvRdd)
+      .csvRdd(sqlContext, csvRdd)
       .collect()
 
     assert(df(0).toSeq === Seq("20", "1.8"))
@@ -483,29 +495,29 @@ abstract class AbstractCsvSuite extends FunSuite {
   }
 
   test("Inserting into csvRdd should throw exception"){
-    val csvRdd = TestSQLContext.sparkContext.parallelize(Seq("age,height", "20,1.8", "16,1.7"))
-    val sampleData = TestSQLContext.sparkContext.parallelize(Seq("age,height", "20,1.8", "16,1.7"))
+    val csvRdd = sqlContext.sparkContext.parallelize(Seq("age,height", "20,1.8", "16,1.7"))
+    val sampleData = sqlContext.sparkContext.parallelize(Seq("age,height", "20,1.8", "16,1.7"))
 
     val df = new CsvParser()
       .withUseHeader(true)
       .withParserLib(parserLib)
-      .csvRdd(TestSQLContext, csvRdd)
+      .csvRdd(sqlContext, csvRdd)
     val sampleDf = new CsvParser()
       .withUseHeader(true)
       .withParserLib(parserLib)
-      .csvRdd(TestSQLContext, sampleData)
+      .csvRdd(sqlContext, sampleData)
 
     df.registerTempTable("csvRdd")
     sampleDf.registerTempTable("sampleDf")
 
     val exception = intercept[java.io.IOException] {
-      sql("INSERT OVERWRITE TABLE csvRdd select * from sampleDf")
+      sqlContext.sql("INSERT OVERWRITE TABLE csvRdd select * from sampleDf")
     }
     assert(exception.getMessage.contains("Cannot INSERT into table with no path defined"))
   }
 
   test("DSL tsv test") {
-    val results = TestSQLContext
+    val results = sqlContext
       .tsvFile(carsTsvFile, parserLib = parserLib)
       .select("year")
       .collect()

--- a/src/test/scala/com/databricks/spark/csv/CsvSuite.scala
+++ b/src/test/scala/com/databricks/spark/csv/CsvSuite.scala
@@ -120,6 +120,8 @@ abstract class AbstractCsvSuite extends FunSuite with BeforeAndAfterAll {
   }
 
   test("DDL test parsing decimal type") {
+    assume(org.apache.spark.SPARK_VERSION.take(3) > "1.3",
+      "DecimalType is broken on Spark 1.3.x")
     sqlContext.sql(
       s"""
          |CREATE TEMPORARY TABLE carsTable

--- a/src/test/scala/com/databricks/spark/csv/util/TextFileSuite.scala
+++ b/src/test/scala/com/databricks/spark/csv/util/TextFileSuite.scala
@@ -17,10 +17,10 @@ package com.databricks.spark.csv.util
 
 import java.nio.charset.UnsupportedCharsetException
 
-import org.apache.spark.sql.test.TestSQLContext
-import org.scalatest.FunSuite
+import org.apache.spark.SparkContext
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
 
-class TextFileSuite extends FunSuite {
+class TextFileSuite extends FunSuite with BeforeAndAfterAll {
   val carsFile = "src/test/resources/cars.csv"
   val carsFile8859 = "src/test/resources/cars_iso-8859-1.csv"
   val numLines = 6
@@ -29,32 +29,47 @@ class TextFileSuite extends FunSuite {
   val utf8 = "utf-8"
   val iso88591 = "iso-8859-1"
 
+  private var sparkContext: SparkContext = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    sparkContext = new SparkContext("local[2]", "TextFileSuite")
+  }
+
+  override def afterAll(): Unit = {
+    try {
+      sparkContext.stop()
+    } finally {
+      super.afterAll()
+    }
+  }
+
   test("read utf-8 encoded file") {
-    val baseRDD = TextFile.withCharset(TestSQLContext.sparkContext, carsFile, utf8)
+    val baseRDD = TextFile.withCharset(sparkContext, carsFile, utf8)
     assert(baseRDD.count() === numLines)
     assert(baseRDD.first().count(_ == ',') == numColumns)
   }
 
   test("read utf-8 encoded file using charset alias") {
-    val baseRDD = TextFile.withCharset(TestSQLContext.sparkContext, carsFile, "utf8")
+    val baseRDD = TextFile.withCharset(sparkContext, carsFile, "utf8")
     assert(baseRDD.count() === numLines)
     assert(baseRDD.first().count(_ == ',') == numColumns)
   }
 
   test("read iso-8859-1 encoded file") {
-    val baseRDD = TextFile.withCharset(TestSQLContext.sparkContext, carsFile8859, iso88591)
+    val baseRDD = TextFile.withCharset(sparkContext, carsFile8859, iso88591)
     assert(baseRDD.count() === numLines)
     assert(baseRDD.first().count(_ == smallThorn) == numColumns)
   }
 
   test("read iso-8859-1 encoded file using charset alias") {
-    val baseRDD = TextFile.withCharset(TestSQLContext.sparkContext, carsFile8859, "8859_1")
+    val baseRDD = TextFile.withCharset(sparkContext, carsFile8859, "8859_1")
     assert(baseRDD.count() === numLines)
     assert(baseRDD.first().count(_ == smallThorn) == numColumns)
   }
 
   test("read iso-8859-1 encoded file with invalid charset") {
-    val baseRDD = TextFile.withCharset(TestSQLContext.sparkContext, carsFile8859, utf8)
+    val baseRDD = TextFile.withCharset(sparkContext, carsFile8859, utf8)
     assert(baseRDD.count() === numLines)
     // file loads but since it's not encoded in utf-8, non-ascii characters are not decoded
     // correctly.
@@ -63,7 +78,7 @@ class TextFileSuite extends FunSuite {
 
   test("unsupported charset") {
     val exception = intercept[UnsupportedCharsetException] {
-      TextFile.withCharset(TestSQLContext.sparkContext, carsFile, "frylock").count()
+      TextFile.withCharset(sparkContext, carsFile, "frylock").count()
     }
     assert(exception.getMessage.contains("frylock"))
   }

